app/(root)/(home)/page.tsx

```diff
@@ -2,33 +2,36 @@ import { getUser } from "@/lib/actions/user.actions";
 import { getAllPosts } from "@/lib/actions/post.actions";
 
 import { CreateType, Post, User } from "@prisma/client";
+import { SearchParams } from "@/types";
 import Posts from "../../../components/post/Posts";
 import Pagination from "@/components/post/Pagination";
 
 export default async function Home({
   searchParams,
 }: {
-  searchParams: { page: string; type: string };
+  searchParams: SearchParams;
 }) {
   const user = (await getUser()) as User;
-
   const posts = await getAllPosts({
     page: searchParams.page ?? "1",
     searchTerm: searchParams.type as CreateType,
+    term: searchParams.tag ?? "",
```

Comment:
In a previous suggestion I suggested you set 'term' to be the value of 'term' that your're setting in your searchbox.
In your `Search.tsx` you have this code:
```js

    const setParams = async () => {
      const newParams = urlManager(searchParams.toString(), {
        page: "1",
        term: searchTerm,
      });
      router.push(`?${newParams}`);
    };
```
So let's actually use that.
```suggestion
    term: searchParams.term ?? "",
```

If you want to get your tag search to work correctly you should use a different variable. "Term" in your URL params should be "term" in your `getAllPosts` function. "tag" in your URL params should be "tag" in you `getAllPosts` function.

The way you currently have it: Searching for a tag does an entire search across the entire post:
```js
// post.actions.ts ... getAllPosts()
...(term && {
          OR: [
            { title: { contains: term, mode: "insensitive" } },
            { content: { contains: term, mode: "insensitive" } },
            { description: { contains: term, mode: "insensitive" } },
            { tags: { has: term } },
          ],
        }),
```

We don't want that. We want the 'tag' search to just search for tags. So you follow the same pattern as before for getting the 'term' to work.

```js
// post.actions.ts  ... getAllPosts()
// adjust the arguments to take in an optional 'tag'
export async function getAllPosts({
  page,
  searchTerm,
  postsToTake = 4,
  term,
  tag,
}: {
  page: string;
  searchTerm?: string;
  postsToTake?: number;
  term?: string;
  tag?: string;
}) {
```
In the 'where' object for our search: add the search for tags only if 'tag' is set.

```js
 where: {
        userEmail,
        ...(searchTerm &&
          searchTerm !== "all" && { createType: searchTerm as CreateType }),
        ...(term && {
          OR: [
            { title: { contains: term, mode: "insensitive" } },
            { content: { contains: term, mode: "insensitive" } },
            { description: { contains: term, mode: "insensitive" } },
            { tags: { has: term } },
          ],
        }),
        ...(tag && { tags: { has: tag } }),
 ```
 
 So if you combine that with my previous suggestion on cleaning up the repetitive code- you get something like this:
 ```js
 export async function getAllPosts({
  page,
  searchTerm,
  postsToTake = 4,
  term,
  tag,
}: {
  page: string;
  searchTerm?: string;
  postsToTake?: number;
  term?: string;
  tag?: string;
}) {
  let hasNextPage = false;
  try {
    const userEmail = await getUserSession();
    // declare the reused options object
    const options = {
      where: {
        userEmail,
        ...(searchTerm &&
          searchTerm !== "all" && { createType: searchTerm as CreateType }),
        ...(term && {
          OR: [
            { title: { contains: term, mode: "insensitive" } },
            { content: { contains: term, mode: "insensitive" } },
            { description: { contains: term, mode: "insensitive" } },
            { tags: { has: term } },
          ],
        }),
        ...(tag && { tags: { has: tag } }),
      },
    };
    
    const somePosts = await prisma.post.findMany({
      ...(options as Prisma.PostFindManyArgs),
      orderBy: [{ createdAt: "desc" }],
      skip: (Number(page) - 1) * postsToTake,
      take: postsToTake + 1,
    });

    const { _count: numberOfResults } = (await prisma.post.aggregate({
      _count: true,
      ...(options as Prisma.PostAggregateArgs),
    })) as any;

    if (somePosts.length > postsToTake) {
      somePosts.pop();
      hasNextPage = true;
    }
    return {
      somePosts,
      hasNextPage,
      numberOfPages: Math.ceil(numberOfResults / postsToTake) || 1,
    };
  } catch (error) {
    console.error("Error returning posts:", error);
    return { error: "An unexpected error occurred while returning posts." };
  }
}
```
And in your homepage your code should look like this in order to actually use the tags/terms/page together:
```js
// app/(root)/(home)/page.tsx
export default async function Home({
  searchParams,
}: {
  searchParams: SearchParams;
}) {
  const user = (await getUser()) as User;
  const posts = await getAllPosts({
    page: searchParams.page ?? "1",
    searchTerm: searchParams.type as CreateType,
    term: searchParams.term ?? "",
    tag: searchParams.tag ?? "",
  });
  ```
  
With all of the suggested changes combined:


https://github.com/DDVVPP/GitNote/assets/4108484/5a7dd52e-f9b7-44af-b760-7386bb57e1bd


Why I'm suggesting these changes based on your current code:

https://www.canva.com/design/DAGB7Qeayxw/13MrjKVFQp2RtxAoDKnAjw/watch?utm_content=DAGB7Qeayxw&utm_campaign=designshare&utm_medium=link&utm_source=editor